pkg/ipam/ippool.go

```diff
@@ -0,0 +1,231 @@
+	size := int(math.Exp2(float64(totalbits-ones))) - 2 // don't count net and broadcast
+	if size < 2 {
+		return nil, errors.New("invalid CIDR Prefix")
+	}
+
+	pool := &IPPool{
+		cidr:      cidr,
+		net:       network,
+		size:      size,
+		available: treemap.NewWithIntComparator(),
+		allocated: make(map[string]bool),
+	}
+	startingIP := ipToInt(pool.net.IP) + 1
+
+	for i := 0; i < pool.size; i++ {
+		intIP := startingIP + i
+		ip := intToIP(startingIP + i).String()
+		pool.available.Put(intIP, ip)
+	}
+
+	// TODO: RecordAvailability(cidr, size)
+
+	return pool, nil
+}
+
+func ipToInt(ip net.IP) int {
+	intIP := ip
+	if len(ip) == 16 {
+		intIP = ip[12:16]
+	}
+
+	return int(binary.BigEndian.Uint32(intIP))
+}
+
+func intToIP(ip int) net.IP {
+	netIP := make(net.IP, 4)
+	binary.BigEndian.PutUint32(netIP, uint32(ip))
+
+	return netIP
+}
+
+func StringIPToInt(stringIP string) int {
+	return ipToInt(net.ParseIP(stringIP))
+}
+
+func (p *IPPool) allocate() (string, error) {
+	p.Lock()
+	defer p.Unlock()
+
+	iter := p.available.Iterator()
+	if !p.available.Empty() {
+		iter.Last()
+		ip := iter.Value().(string)
+		p.allocated[ip] = true
+		p.available.Remove(iter.Key())
+		// TODO: RecordAllocateGlobalIP(p.cidr)
+
+		return ip, nil
+	}
+
+	return "", errors.New("IPAM: Insufficient IPs available for allocation")
+}
+
+func (p *IPPool) Allocate(num int) ([]string, error) {
+	if p.available.Size() < num {
+		return nil, errors.New("IPAM: Insufficient IPs available for allocation")
+	}
+
+	ips := make([]string, num)
+
+	if num == 1 {
+		ip, err := p.allocate()
+		if err != nil {
+			return nil, err
+		}
+		ips[0] = ip
+		return ips, nil
+	}
+
+	intIPs := make([]int, num)
+	prev, current := 0, 0
+
+	p.Lock()
+	defer p.Unlock()
+
+	iter := p.available.Iterator()
+	for iter.Next() {
+		if current == num {
+			return p.allocateBlock(intIPs), nil
+		}
+
+		intIP := iter.Key().(int)
+		intIPs[current] = intIP
+		if current == 0 {
+			current++
+			continue
+		}
+
+		prevIntIP := intIPs[prev]
+		if prevIntIP+1 != intIP {
+			intIPs = make([]int, num)
+			prev, current = 0, 0
+		} else {
+			prev = current
+			current++
+		}
+	}
+
+	return nil, errors.New("IPAM: unable to find contiguous block for allocation")
+}
+
+func (p *IPPool) Release(ips []string) {
+	p.Lock()
+	defer p.Unlock()
+
+	for _, ip := range ips {
+		p.available.Put(StringIPToInt(ip), ip)
+		delete(p.allocated, ip)
+	}
+	p.Unlock()
+}
+
+func (p *IPPool) Reserve(ips []string) error {
+	num := len(ips)
+	intIPs := make([]int, num)
+	p.Lock()
+	defer p.Unlock()
+
+	if p.available.Size() < num {
+		errors.New("IPAM: Insufficient IPs available for allocation")
+	}
+
+	for i := 0; i < num; i++ {
+		intIPs[i] = StringIPToInt(ips[i])
+		_, found := p.available.Get(intIPs[i]); if !found {
+			return errors.New(fmt.Sprintf("IPAM: Requested IP %s already allocated", ips[i]))
+		}
+	}
+
+	for i := 0; i < num; i++ {
+		p.allocated[ips[i]] = true
+		p.available.Remove(intIPs[i])
```

Comment:
I think we can move this after L183 instead of a second loop?

Reply:
No. First we check if requested no. of IPs can be allocated or not. Idea is to fail if even one of the requested is not available. Unless you want to support use case of partial allocation. If we move this to after L83, we will have to add another loop for rollback of ones that were allocated.